### PR TITLE
[C++] Add concepts for v2's reverse-string

### PR DESCRIPTION
**Pull request by AlexLeSang**

### CODE REVIEW

File: languages/cpp/reference/exercise-concepts/reverse-string.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+}
+```
+
+## Concepts
+
+- .h file: the tests include `"reverse_string.h"`
+- .cpp file:  the most common approach is to use a `.cpp` for the definitions
+- namespaces: the function is member of a namespace
+- functions: used as the main entry point for the exercise
+- function arguments: the function has two arguments
```
Comment (marvelou-s) [inline languages/cpp/reference/exercise-concepts/reverse-string.md]: Nit. There're no two arguments in function ;)

File: languages/cpp/reference/exercise-concepts/reverse-string.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+- strings: input is a string (or maybe string_view)
+- return values: the result is returned by the function
+- pass by reference to const: avoid making copy of input
```
Comment (marvelou-s) [inline languages/cpp/reference/exercise-concepts/reverse-string.md]: References should be an independent concept (in addition to that)

File: languages/cpp/reference/exercise-concepts/reverse-string.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+- iterators: view input as a range
+
+
+## Alternative approaches
+
```
Comment (marvelou-s) [inline languages/cpp/reference/exercise-concepts/reverse-string.md]: `std::reverse()` from algorithm library looks like good alternative approach.

File: languages/cpp/reference/exercise-concepts/reverse-string.md [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+`reverse_string.cpp`
+```cpp
+#include "reverse_string.h"
+
+namespace reverse_string
+{
+
+std::string reverse_string(const std::string& text)
+{
+    return std::string(text.rbegin(), text.rend());
```
Comment (marvelou-s) [inline languages/cpp/reference/exercise-concepts/reverse-string.md]: Could the constructor be a concept by itself? I tend to think yes.

File: languages/cpp/reference/exercise-concepts/reverse-string.md (hunk 3 above)
Comment (AlexLeSang) [reply languages/cpp/reference/exercise-concepts/reverse-string.md]: In this case, i do not think so. It requires additional copy to be constructed before you can call it. Also, according to the compiler explorer, it generates more code on gcc-trunk.

File: languages/cpp/reference/exercise-concepts/reverse-string.md (hunk 1 above)
Comment (AlexLeSang) [reply languages/cpp/reference/exercise-concepts/reverse-string.md]: Yes, thank you, I have missed that =)

File: languages/cpp/reference/exercise-concepts/reverse-string.md (hunk 2 above)
Comment (AlexLeSang) [reply languages/cpp/reference/exercise-concepts/reverse-string.md]: Agree, I'll add that.

File: languages/cpp/reference/exercise-concepts/reverse-string.md (hunk 4 above)
Comment (AlexLeSang) [reply languages/cpp/reference/exercise-concepts/reverse-string.md]: I agree, that is correct observation!

File: languages/cpp/reference/exercise-concepts/reverse-string.md (hunk 3 above)
Comment (marvelou-s) [reply languages/cpp/reference/exercise-concepts/reverse-string.md]: i fully agree. it's not the most **effective** way to solve it. But there are a lot of approved solutions on `exercism` used this approach.

Comment (silvanocerza) [reply languages/cpp/reference/exercise-concepts/reverse-string.md]: I agree with @marvelou-s here, alternative approaches should not take into consideration efficiency too much.
If someone submitted a solution that's not the most efficient we should take it into consideration anyway.

Comment (AlexLeSang) [reply languages/cpp/reference/exercise-concepts/reverse-string.md]: I see your point, I'll add that to the list of alternative approaches.

**Pull request merged**